File: cpp/21-MergeTwoSortedLists/main.cpp
#include <iostream>

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution 
{
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) 
    {
        if(!list1) return list2;
        if(!list2) return list1;

        auto head = grab(list1, list2);
        auto next = head;

        while(list1 && list2)
        {
            next->next = grab(list1, list2);
            next = next->next;
        }

        next->next = (list1 ? list1 : list2);

        return head;
    }

    ListNode *grab(ListNode *&list1, ListNode *&list2)
    {
        ListNode *next = nullptr;
        if(list1->val < list2->val)
        {
            next = list1;
            list1 = list1->next;
        }
        else
        {
            next = list2;
            list2 = list2->next;
        }

        return next;
    }
};

void print(ListNode *node)
{
    if(node == nullptr)
    {
        std::cout << "null";
    }
    else
    {
        while(node)
        {
            std::cout << node->val << ' ';
            node = node->next;
        }
    }

    std::cout << "\n";
}

void test1()
{
    auto list1 = new ListNode(1, new ListNode(2, new ListNode(4)));
    auto list2 = new ListNode(1, new ListNode(3, new ListNode(4)));

    Solution solution;
    print(solution.mergeTwoLists(list1, list2));
}

void test2()
{
    ListNode *list1 = nullptr;
    ListNode *list2 = nullptr;

    Solution solution;
    print(solution.mergeTwoLists(list1, list2));
}

void test3()
{
    ListNode *list1 = nullptr;
    auto list2 = new ListNode(0);

    Solution solution;
    print(solution.mergeTwoLists(list1, list2));
}

int main()
{
    test1();
    test2();
    test3();
}

File: go/21-MergeTwoSortedLists/main.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

func grabNext(list1 *ListNode, list2 *ListNode) (*ListNode, *ListNode, *ListNode) {
	if list1.Val < list2.Val {
		return list1, list1.Next, list2
	} else {
		return list2, list1, list2.Next
	}
}

func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {
	if list1 == nil {
		return list2
	}

	if list2 == nil {
		return list1
	}

	next, l1, l2 := grabNext(list1, list2)
	head := next

	for l1 != nil && l2 != nil {
		next.Next, l1, l2 = grabNext(l1, l2)
		next = next.Next
	}

	if l1 != nil {
		next.Next = l1
	} else {
		next.Next = l2
	}

	return head
}

func printList(list *ListNode) {
	for list != nil {
		fmt.Print(list.Val)
		list = list.Next
	}

	fmt.Println()
}

func test1() {
	list1 := &ListNode{Val: 1, Next: &ListNode{Val: 2, Next: &ListNode{Val: 4}}}
	list2 := &ListNode{Val: 1, Next: &ListNode{Val: 3, Next: &ListNode{Val: 4}}}

	printList(mergeTwoLists(list1, list2))
}

func main() {
	test1()
}
